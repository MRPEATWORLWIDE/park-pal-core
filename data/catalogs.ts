import catalogsJson from './json/catalogs.json'
import taxonomiesJson from './json/taxonomies.json'
import taxonsJson from './json/taxons.json'

type JsonCatalog = {
  key: string
  name: string
  taxonomies: string[]
}

type JsonTaxonomy = {
  key: string
  label?: string
  name: string
  taxons: string[]
}

export type Catalog = Omit<JsonCatalog, 'taxonomies'> & {
  taxonomies: Taxonomy[]
}

export type Taxonomy = Omit<JsonTaxonomy, 'taxons'> & {
  taxons: Taxon[]
}

export type Taxon = {
  key: string
  label: string
  description: string
  name: string
  slug: string
  image: string
  references: string[]
}

const catalogs: JsonCatalog[] = catalogsJson
const taxonomies: JsonTaxonomy[] = taxonomiesJson
const taxons: Taxon[] = taxonsJson

const resolveCatalogs = (): Catalog[] => {
  const resolvedTaxonomies = resolveTaxonomies()
  return catalogs.map(catalog => {
    return {
      ...catalog,
      taxonomies: catalog.taxonomies.map(taxonomyKey => resolvedTaxonomies.find(taxonomy => taxonomy.key === taxonomyKey)!)
    }
  })
}

const resolveTaxonomies = (): Taxonomy[] => {
  return taxonomies.map(taxonomy => {
    return {
      ...taxonomy,
      taxons: taxonomy.taxons.map(taxonKey => taxons.find(taxon => taxon.key === taxonKey)!)
    }
  })
}

export const getCatalog = (name: string): Catalog => {
  const catalog = resolveCatalogs().find(catalog => catalog.name === name)

  if (!catalog) {
    throw new Error(`Cannot find the catalog with name "${name}"`)
  }

  return catalog
}