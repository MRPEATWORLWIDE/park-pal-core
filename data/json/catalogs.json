[
  {
    "key": "catalog_1",
    "name": "AMER",
    "taxonomies": ["taxonomy_1", "taxonomy_4"]
  },
  {
    "key": "catalog_2",
    "name": "EMEA",
    "taxonomies": ["taxonomy_2", "taxonomy_4"]
  },
  {
    "key": "catalog_3",
    "name": "APAC",
    "taxonomies": ["taxonomy_3", "taxonomy_5"]
  }
]