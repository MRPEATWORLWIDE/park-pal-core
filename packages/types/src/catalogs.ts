import { z } from 'zod'
import { localizedFieldSchema } from './locale'

const catalog_schema = z.object({
  id: z.string(),
  name: z.string(),
  taxonomies: z.string().array()
})

const taxonomy_schema = z.object({
  id: z.string(),
  name: z.string(),
  facetKey: z.string(),
  taxons: z.string().array()
})

const taxon_schema = z.object({
  id: z.string(),
  name: z.string(),
  label: localizedFieldSchema(z.string()),
  description: localizedFieldSchema(z.string()),
  slug: z.string().transform(slug => slug.replace(/^\//, '')),
  references: z.string().array(),
  taxons: z.string().array().optional()
})

export const rawDataCatalogs_schema = catalog_schema.array()
export const rawDataTaxonomies_schema = taxonomy_schema.array()
export const rawDataTaxons_schema = taxon_schema.array()

export type RawDataCatalog = z.infer<typeof catalog_schema>
export type RawDataTaxonomy = z.infer<typeof taxonomy_schema>
export type RawDataTaxon = z.infer<typeof taxon_schema>
