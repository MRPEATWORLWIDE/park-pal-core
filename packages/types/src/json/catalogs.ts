import { z } from 'zod'

const catalog_schema = z.object({
  id: z.string(),
  name: z.string(),
  taxonomies: z.string().array()
})

export const rawDataCatalogs_schema = catalog_schema.array()

export type RawDataCatalog = z.infer<typeof catalog_schema>
