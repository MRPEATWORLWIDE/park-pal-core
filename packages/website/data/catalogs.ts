import { fetchJsonData } from '#utils/data'
import { memoize } from '#utils/memoize'
import { makeUnserializable, type Unserializable } from '#utils/unserializable'
import { type RawDataCatalog, rawDataCatalogs_schema } from '@commercelayer/demo-store-types'

export const getRawDataCatalogs = memoize(
  async function (): Promise<Unserializable<RawDataCatalog[]>> {
    return makeUnserializable(
      rawDataCatalogs_schema.parse(
        await fetchJsonData('catalogs')
      )
    )
  }
)
