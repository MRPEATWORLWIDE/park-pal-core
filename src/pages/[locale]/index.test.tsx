import { Catalog, Taxon, Taxonomy } from '#data/catalogs'
import { render, screen, within } from '@testing-library/react'
import { I18nProvider } from 'next-localization'

import Home from './index.page'

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: {
        locale: 'en-US'
      },
      asPath: '',
    };
  },
}));

test('home', () => {
  const taxon: Taxon = {
    key: 'taxon_1',
    label: 'Accessories',
    description: 'Accessories',
    slug: 'accessories',
    name: 'All Accessories',
    image: '/assets/images/samsung-case-samsung-galaxy-s21-plus-lifestyle-4-60f836d4aaa9d.png',
    references: [
      'APRONXXX000000FFFFFFXXXX'
    ]
  }

  const taxonomy: Taxonomy = {
    key: 'taxonomy_1',
    label: 'Shop by categories',
    name: 'Default Category',
    taxons: [taxon]
  }

  const catalog: Catalog = {
    key: 'catalog_1',
    name: 'AMER',
    taxonomies: [taxonomy]
  }

  const { container } = render(
    <I18nProvider lngDict={{ general: { title: 'Welcome to' } }} locale='en'>
      <Home catalog={catalog} />
    </I18nProvider>
  )

  const main = within(screen.getByRole('main'))

  expect(
    main.getByRole('heading', { level: 2, name: /Shop by categories/i })
  ).toBeDefined()

  expect(container).toMatchSnapshot()
})
