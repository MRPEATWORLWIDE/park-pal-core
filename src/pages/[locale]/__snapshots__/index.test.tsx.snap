// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home 1`] = `
<div>
  <main>
    <div
      class="container mx-auto px-4"
    >
      <nav
        class="flex items-center justify-between flex-wrap"
      >
        <div
          class="flex items-center flex-no-shrink text-white mr-6"
        >
          <a
            href="/"
          >
            <div
              class="py-6 pb-6 "
            >
              <img
                alt="Commerce Layer logomark"
                class="inline"
                src="https://data.commercelayer.app/assets/logos/glyph/black/commercelayer_glyph_black.svg"
                width="32"
              />
              <img
                alt="Commerce Layer logotype"
                class="ml-4 hidden md:inline"
                src="https://data.commercelayer.app/assets/logos/logotype/black/commercelayer_logotype_black.svg"
                width="192"
              />
            </div>
          </a>
        </div>
        <div
          class="flex items-center w-auto flex-grow justify-end"
        >
          <a
            class="block lg:inline-block mr-4"
          >
            <svg />
          </a>
          <a
            class="block lg:inline-block mr-4"
          >
            <svg />
          </a>
          <a
            class="block lg:inline-block"
            href="/cart"
          >
            <svg />
          </a>
        </div>
      </nav>
      <div>
        <p>
          <a
            href="/"
          >
            Country Selector
          </a>
            |  
          <a
            href="/?locale="
          >
            English
          </a>
            |  
          <a
            href="/?locale="
          >
            Italiano
          </a>
            |  
          <a
            href="/search"
          >
            All Products
          </a>
        </p>
      </div>
      <div>
        <h2
          class="mt-16 block text-2xl font-semibold text-black"
        >
          Shop by categories
        </h2>
        <div
          class="mt-6 space-y-12 md:space-y-0 md:grid md:grid-cols-2 lg:grid-cols-3 md:gap-6"
        >
          <a
            class="relative w-full h-80 xl:h-96 bg-white rounded-lg overflow-hidden group-hover:opacity-75 sm:aspect-w-2 sm:aspect-h-1 lg:aspect-w-1 lg:aspect-h-1"
            href="/search/accessories"
          >
            <img
              alt="Accessories"
              class="object-cover w-full h-full"
              src="/demo-store/assets/images/samsung-case-samsung-galaxy-s21-plus-lifestyle-4-60f836d4aaa9d.png"
            />
            <div
              class="absolute inset-0 bg-gradient-to-b from-black to-black/0 opacity-50"
            />
            <div
              class="absolute p-4 top-0 inset-x-0 text-white leading-snug"
            >
              <h3
                class="text-lg font-semibold"
              >
                Accessories
              </h3>
              <div
                class="text-md font-medium text-gray-300"
              >
                Accessories
              </div>
            </div>
            <div
              class="absolute p-4 bottom-0 inset-x-0 text-white leading-snug text-lg"
            >
              <div
                class="font-semibold"
              >
                View all 
                -&gt;
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </main>
</div>
`;
